File: lib/tauri/filesystem.ts
import { invoke } from '@tauri-apps/api/core';

export interface FileInfo {
  name: string;
  path: string;
  size: number;
  is_dir: boolean;
  modified: string;
}

/**
 * Filesystem management functions
 */
export const filesystemCommands = {
  /**
   * Read text file contents
   */
  readFile: async (path: string): Promise<string> => {
    try {
      return await invoke<string>('read_file', { path });
    } catch (error) {
      throw new Error(`Failed to read file ${path}: ${error}`);
    }
  },

  /**
   * Read binary file contents
   */
  readFileBinary: async (path: string): Promise<Uint8Array> => {
    try {
      const data = await invoke<number[]>('read_file_binary', { path });
      return new Uint8Array(data);
    } catch (error) {
      throw new Error(`Failed to read binary file ${path}: ${error}`);
    }
  },

  /**
   * Write text file contents
   */
  writeFile: async (path: string, content: string): Promise<void> => {
    try {
      await invoke('write_file', { path, content });
    } catch (error) {
      throw new Error(`Failed to write file ${path}: ${error}`);
    }
  },

  /**
   * Write binary file contents
   */
  writeFileBinary: async (path: string, content: Uint8Array): Promise<void> => {
    try {
      const data = Array.from(content);
      await invoke('write_file_binary', { path, content: data });
    } catch (error) {
      throw new Error(`Failed to write binary file ${path}: ${error}`);
    }
  },

  /**
   * Delete file
   */
  deleteFile: async (path: string): Promise<void> => {
    try {
      await invoke('delete_file', { path });
    } catch (error) {
      throw new Error(`Failed to delete file ${path}: ${error}`);
    }
  },

  /**
   * Create directory (recursively creates parent directories)
   */
  createDirectory: async (path: string): Promise<void> => {
    try {
      await invoke('create_directory', { path });
    } catch (error) {
      throw new Error(`Failed to create directory ${path}: ${error}`);
    }
  },

  /**
   * Delete directory (recursively deletes all contents)
   */
  deleteDirectory: async (path: string): Promise<void> => {
    try {
      await invoke('delete_directory', { path });
    } catch (error) {
      throw new Error(`Failed to delete directory ${path}: ${error}`);
    }
  },

  /**
   * List directory contents
   */
  listDirectory: async (path: string): Promise<FileInfo[]> => {
    try {
      return await invoke<FileInfo[]>('list_directory', { path });
    } catch (error) {
      throw new Error(`Failed to list directory ${path}: ${error}`);
    }
  },

  /**
   * Check if file exists
   */
  fileExists: async (path: string): Promise<boolean> => {
    try {
      return await invoke<boolean>('file_exists', { path });
    } catch (error) {
      throw new Error(`Failed to check if file exists ${path}: ${error}`);
    }
  },

  /**
   * Get file information
   */
  getFileInfo: async (path: string): Promise<FileInfo> => {
    try {
      return await invoke<FileInfo>('get_file_info', { path });
    } catch (error) {
      throw new Error(`Failed to get file info for ${path}: ${error}`);
    }
  },

  /**
   * Copy file
   */
  copyFile: async (source: string, destination: string): Promise<void> => {
    try {
      await invoke('copy_file', { source, destination });
    } catch (error) {
      throw new Error(`Failed to copy file from ${source} to ${destination}: ${error}`);
    }
  },

  /**
   * Move file
   */
  moveFile: async (source: string, destination: string): Promise<void> => {
    try {
      await invoke('move_file', { source, destination });
    } catch (error) {
      throw new Error(`Failed to move file from ${source} to ${destination}: ${error}`);
    }
  },
};

File: src-tauri/src/commands/filesystem.rs
use std::path::PathBuf;
use std::fs;

#[derive(Debug, serde::Serialize, serde::Deserialize)]
pub struct FileInfo {
    pub name: String,
    pub path: String,
    pub size: u64,
    pub is_dir: bool,
    pub modified: String,
}

#[tauri::command]
pub async fn read_file(path: String) -> Result<String, String> {
    fs::read_to_string(&path)
        .map_err(|e| format!("Failed to read file {}: {}", path, e))
}

#[tauri::command]
pub async fn read_file_binary(path: String) -> Result<Vec<u8>, String> {
    fs::read(&path)
        .map_err(|e| format!("Failed to read file {}: {}", path, e))
}

#[tauri::command]
pub async fn write_file(path: String, content: String) -> Result<(), String> {
    // Create parent directories if they don't exist
    if let Some(parent) = PathBuf::from(&path).parent() {
        fs::create_dir_all(parent)
            .map_err(|e| format!("Failed to create directories: {}", e))?;
    }
    
    fs::write(&path, content)
        .map_err(|e| format!("Failed to write file {}: {}", path, e))
}

#[tauri::command]
pub async fn write_file_binary(path: String, content: Vec<u8>) -> Result<(), String> {
    // Create parent directories if they don't exist
    if let Some(parent) = PathBuf::from(&path).parent() {
        fs::create_dir_all(parent)
            .map_err(|e| format!("Failed to create directories: {}", e))?;
    }
    
    fs::write(&path, content)
        .map_err(|e| format!("Failed to write file {}: {}", path, e))
}

#[tauri::command]
pub async fn delete_file(path: String) -> Result<(), String> {
    fs::remove_file(&path)
        .map_err(|e| format!("Failed to delete file {}: {}", path, e))
}

#[tauri::command]
pub async fn create_directory(path: String) -> Result<(), String> {
    fs::create_dir_all(&path)
        .map_err(|e| format!("Failed to create directory {}: {}", path, e))
}

#[tauri::command]
pub async fn delete_directory(path: String) -> Result<(), String> {
    fs::remove_dir_all(&path)
        .map_err(|e| format!("Failed to delete directory {}: {}", path, e))
}

#[tauri::command]
pub async fn list_directory(path: String) -> Result<Vec<FileInfo>, String> {
    let entries = fs::read_dir(&path)
        .map_err(|e| format!("Failed to read directory {}: {}", path, e))?;
    
    let mut files = Vec::new();
    
    for entry in entries {
        let entry = entry.map_err(|e| format!("Failed to read directory entry: {}", e))?;
        let metadata = entry.metadata()
            .map_err(|e| format!("Failed to read metadata: {}", e))?;
        
        let modified = metadata.modified()
            .map(|time| {
                use std::time::SystemTime;
                let duration = time.duration_since(SystemTime::UNIX_EPOCH).unwrap();
                chrono::DateTime::from_timestamp(duration.as_secs() as i64, 0)
                    .map(|dt| dt.to_rfc3339())
                    .unwrap_or_else(|| "Unknown".to_string())
            })
            .unwrap_or_else(|_| "Unknown".to_string());
        
        files.push(FileInfo {
            name: entry.file_name().to_string_lossy().to_string(),
            path: entry.path().to_string_lossy().to_string(),
            size: metadata.len(),
            is_dir: metadata.is_dir(),
            modified,
        });
    }
    
    Ok(files)
}

#[tauri::command]
pub async fn file_exists(path: String) -> Result<bool, String> {
    Ok(PathBuf::from(path).exists())
}

#[tauri::command]
pub async fn get_file_info(path: String) -> Result<FileInfo, String> {
    let path_buf = PathBuf::from(&path);
    
    if !path_buf.exists() {
        return Err(format!("File does not exist: {}", path));
    }
    
    let metadata = fs::metadata(&path)
        .map_err(|e| format!("Failed to read metadata: {}", e))?;
    
    let modified = metadata.modified()
        .map(|time| {
            use std::time::SystemTime;
            let duration = time.duration_since(SystemTime::UNIX_EPOCH).unwrap();
            chrono::DateTime::from_timestamp(duration.as_secs() as i64, 0)
                .map(|dt| dt.to_rfc3339())
                .unwrap_or_else(|| "Unknown".to_string())
        })
        .unwrap_or_else(|_| "Unknown".to_string());
    
    Ok(FileInfo {
        name: path_buf.file_name()
            .map(|n| n.to_string_lossy().to_string())
            .unwrap_or_else(|| "Unknown".to_string()),
        path: path.clone(),
        size: metadata.len(),
        is_dir: metadata.is_dir(),
        modified,
    })
}

#[tauri::command]
pub async fn copy_file(source: String, destination: String) -> Result<(), String> {
    // Create parent directories if they don't exist
    if let Some(parent) = PathBuf::from(&destination).parent() {
        fs::create_dir_all(parent)
            .map_err(|e| format!("Failed to create directories: {}", e))?;
    }
    
    fs::copy(&source, &destination)
        .map_err(|e| format!("Failed to copy file from {} to {}: {}", source, destination, e))?;
    
    Ok(())
}

#[tauri::command]
pub async fn move_file(source: String, destination: String) -> Result<(), String> {
    // Create parent directories if they don't exist
    if let Some(parent) = PathBuf::from(&destination).parent() {
        fs::create_dir_all(parent)
            .map_err(|e| format!("Failed to create directories: {}", e))?;
    }
    
    fs::rename(&source, &destination)
        .map_err(|e| format!("Failed to move file from {} to {}: {}", source, destination, e))?;
    
    Ok(())
}
